src/socks-to-rtc/socks-to-rtc.ts

```
@@ -65,92 +83,79 @@ module SocksToRtc {
+    public start = (
+        tcpServer:Tcp.Server,
+        peerconnection:freedom_UproxyPeerConnection.Pc)
+        : Promise<Net.Endpoint> => {
+      if (this.tcpServer_) {
+        throw new Error('already configured');
       }
-      this.isStopped_ = true;
-      this.signalsForPeer.clear();
-      this.bytesReceivedFromPeer.clear();
-      this.bytesSentToPeer.clear();
-      this.peerConnection_.close();
-      this.sessions_ = {};
-      return this.tcpServer_.shutdown();
-    }
+      this.tcpServer_ = tcpServer;
+      this.peerConnection_ = peerconnection;
 
-    private setupPeerConnection_ = (
-        pcConfig:WebRtc.PeerConnectionConfig,
-        obfuscate?:boolean)
-        : Promise<WebRtc.ConnectionAddresses> => {
-      // SOCKS sessions biject to peerconnection datachannels.
-      this.peerConnection_ = obfuscate ?
-          freedom.churn(pcConfig) :
-          freedom['core.uproxypeerconnection'](pcConfig);
       this.peerConnection_.on('dataFromPeer', this.onDataFromPeer_);
-      this.peerConnection_.on('peerOpenedChannel', (channelLabel:string) => {
-        log.error('unexpected peerOpenedChannel event: ' +
-            JSON.stringify(channelLabel));
-      });
-      this.peerConnection_.on('signalForPeer',
-          this.signalsForPeer.handle);
-
-      var onceConnected = this.peerConnection_.onceConnected()
-      this.peerConnection_.negotiateConnection();
-      // Give back onceConnected endpoint, but only after a control channel has
-      // been setupp.
-      return onceConnected.then(() => {
-          return this.peerConnection_.openDataChannel('_control_')
-        })
-        .then(() => {
-          this.peerConnection_.send('_control_', { str: 'hello?' });
-          return onceConnected;
+      this.peerConnection_.on('signalForPeer', this.signalsForPeer.handle);
+
+      // Start and listen for notifications.
+      peerconnection.negotiateConnection();
+      this.onceReady = Promise.all<any>([
```

I've added a TODO to use `tcpServer.onceListening` once it's available. I would prefer that too.

Regarding `Promise<any>`, what you suggest isn't _exactly_ the same as `Promise.all` because it won't fail fast. What I can do is ignore the returned array and instead return `tcpServer.endpoint`.